File: 240628/미래가 보이는 가위바위보/rock-paper-scissors-to-see-the-future.py
N = int(input())
B = [input() for _ in range(N)]

def fight(a,b):
    if a == "H" and b == "S":
        return 1
    elif a == "S" and b == "P":
        return 1
    elif a == "P" and b == "H":
        return 1
    return 0

L = [[0] * (N+1) for _ in range(3)]
R = [[0] * (N+1) for _ in range(3)]

for i in range(1, N+1):
    L[0][i] = L[0][i-1] + fight("H", B[i-1])
    L[1][i] = L[1][i-1] + fight("S", B[i-1])
    L[2][i] = L[2][i-1] + fight("P", B[i-1])

for i in range(N-1, -1, -1):
    R[0][i] = R[0][i+1] + fight("H", B[i])
    R[1][i] = R[1][i+1] + fight("S", B[i])
    R[2][i] = R[2][i+1] + fight("P", B[i])    

ans =0 
for i in range(N+1):
    lm = max(L[0][i], L[1][i], L[2][i])
    rm = max(R[0][i], R[1][i], R[2][i])
    ans = max(ans, lm + rm)
print(ans)

File: 240903/미래가 보이는 가위바위보/rock-paper-scissors-to-see-the-future.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();

        char[] B = new char[n];

        for (int i = 0; i < n; i++){
            String k = sc.next();
            B[i] = k.charAt(0);
        }

        int[] H = new int[n];
        int[] S = new int[n];
        int[] P = new int[n];

        int[] LH = new int[n];
        int[] LS = new int[n];
        int[] LP = new int[n];

        int[] RH = new int[n];
        int[] RS = new int[n];
        int[] RP = new int[n];

        for (int i = 0; i < n; i++){
            if (B[i] == 'H')
                P[i] = 1;
            if (B[i] == 'S')
                H[i] = 1;
            if (B[i] == 'P')
                S[i] = 1;
        }

        LH[0] = H[0];
        LS[0] = S[0];
        LP[0] = P[0];

        RH[n-1] = H[n-1];
        RS[n-1] = S[n-1];
        RP[n-1] = P[n-1];

        for (int i = 1; i < n; i++){
            LH[i] = LH[i-1] + H[i];
            LS[i] = LS[i-1] + S[i];
            LP[i] = LP[i-1] + P[i];
        }

        for (int i = n-2; i >= 0; i--){
            RH[i] = RH[i+1] + H[i];
            RS[i] = RS[i+1] + S[i];
            RP[i] = RP[i+1] + P[i];
        }

        int ans = LH[n-1];
        ans = Math.max(ans, LS[n-1]);
        ans = Math.max(ans, LP[n-1]);

        for (int i = 0; i < n - 1; i++){
            int r = Math.max(RH[i+1], RS[i+1]);
            r = Math.max(r, RP[i+1]);
            
            int l = Math.max(LH[i], LS[i]);
            l = Math.max(l, LP[i]);

            ans = Math.max(ans, l + r);
        }
        System.out.println(ans);
    }
}